fog/overseer/server/src/worker.rs

```diff
@@ -0,0 +1,544 @@
+    /// for an outstanding ingress key, if it exists.
+    ///
+    /// The logic is as follows:
+    ///   1. Find all of the "outstanding keys" as determined by the RecoveryDb.
+    ///      These are ingress keys that Fog Ingest needs to scan blocks with
+    ///      but Fog Ingest isn't currently doing that because all nodes are
+    ///      idle.
+    ///   2. If there are:
+    ///         a) 0 outstanding keys:
+    ///              No node will be activated. We now have to:
+    ///                 (i)  Set new keys on an idle node.
+    ///                 (ii) Activate that node.
+    ///         b) 1 oustanding key:
+    ///              Try to find an idle node that contains that key.
+    ///                 (i)  If you find one, great! Just activate that node. If
+    ///                      activation is unsuccessful, then return an error
+    ///                      and return to the overseer polling logic.
+    ///                 (ii) If you don't find an idle node with that key,
+    ///                      then you have to report that key as lost.
+    ///        c) > 1 outstanding key:
+    ///             (i) Disable
+    ///             (ii) TODO: Send an alert.
+    fn perform_automatic_failover(
+        &self,
+        ingest_summary_node_mappings: Vec<IngestSummaryNodeMapping>,
+    ) -> Result<(), OverseerError> {
+        let inactive_outstanding_keys: Vec<CompressedRistrettoPublic> =
+            self.get_inactive_outstanding_keys()?;
+
+        match inactive_outstanding_keys.len() {
+            0 => log::info!(self.logger, "Found 0 outstanding keys."),
+            1 => log::info!(self.logger, "Found 1 outstanding key."),
+            _ => {
+                log::error!(self.logger, "Found multiple outstanding keys. This requires manual intervention. Disabling overseer.");
+                self.is_enabled.store(false, Ordering::SeqCst);
+                return Err(OverseerError::MultipleOutstandingKeys("Multiple outstanding keys found. This is unexpected and requires manual intervention. As such, we've disabled overseer. Take action and then enable overseer.".to_string()));
+            }
+        }
+
+        if let Some(inactive_outstanding_key) = inactive_outstanding_keys.get(0) {
+            match self.try_to_activate_node_for_inactive_outstanding_key(
+                inactive_outstanding_key,
+                &ingest_summary_node_mappings,
+            ) {
+                Ok(_) => {
+                    log::info!(
+                        self.logger,
+                        "A node was activated, so there is no need to report the key as lost."
+                    );
+                    return Ok(());
+                }
+                Err(err) => log::error!(
+                    self.logger,
+                    "There was an issue activating a node for the outstanding key: {:?}",
+                    err
+                ),
```

Comment:
i think perhaps we should not fall through here?

Reply:
Agreed. Changed.